doc/blog: announce and introduce cli execution (#1138)
ariga/atlas: doc/website/blog/2022-09-05-announcing-migration-execution.mdx

```

</TabItem>

<TabItem value="sum">

```text
h1:5zAQ8J0qziTKWg09fRNdbUf3rnLdvA1KHayh8l1SgM0=
20220908105652.sql h1:4WEB39tqALVYNQJTfULKizxEuUC37sgFs0LN5dKJpOw=

```

</TabItem>

</Tabs>

Now, let's create a new migration file to create a table `tbl_3` and update the directory integrity file.

```shell
atlas migrate new add_table --dir "file://migrations"
echo "CREATE TABLE `tbl_3` (`col` text NULL);" >> migrations/$(ls -t migrations | head -n1)
atlas migrate hash --force --dir "file://migrations"
```

Since we now have both a migration file representing our current database state and the new migration file to apply,
we can make use of the `--baseline` flag:

```shell
atlas migrate apply \
  --dir "file://migrations" \
  --url "mysql://root:pass@localhost:3306/my_schema" \
  --baseline "20220908110527" # replace the version with the one generated by you
```

```shell
Migrating to version 20220908110847 from 20220908110527 (1 migrations in total):

  -- migrating version 20220908110847
    -> CREATE TABLE `tbl_3` (`col` text NULL);
  -- ok (14.325493ms)

  -------------------------
  -- 15.786455ms
  -- 1 migrations
  -- 1 sql statements
```

> ### Outlook

Reviewer: Mention that it is powering Ent.

---
title: The Atlas Migration Execution Engine
authors: masseelch
tags: [atlas, migrations, versioned, announcement]
image: https://blog.ariga.io/uploads/images/posts/v0.6.0/atlas-migrate-diff.png
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

With the release of [v0.6.0](https://github.com/ariga/atlas/releases/tag/v0.6.0), we
[introduced](https://atlasgo.io/blog/2022/08/11/announcing-versioned-migration-authoring) a workflow for managing
changes to database schemas that we have called: _Versioned Migration Authoring_.

Today, we released the first version of the Atlas migration execution engine, that can apply migration files on
your database. In this post, we will give a brief overview of the features and what to expect in the future.

### Migration File Format

At Atlas, we believe in a linear migration history. Therefore, Atlas migration files format does not have the concept
of rollback migrations. To undo a previously applied migration file, simply create a new migration file reverting

> the changes done in the file before.

Reviewer 2: Wrote you privately about this section

> Now, some might argue, that rollback migration exist to have a fallback mechanism in case a migration execution fails.

Reviewer 3: ```suggestion
Now, some might argue, that rollback migrations exist to have a fallback mechanism in case a migration execution fails.
```

> This, however, is deceptive, since

Reviewer 3: ```suggestion
This, however, is deceptive, since:
```

> 1. some SQL statements are simply not revertible, (e.g. destructive changes like

Reviewer 3: ```suggestion
1. Some SQL statements are simply not revertible, (e.g. destructive changes like
```

dropping a column),

> 2. execution order of the rollback migration is not easy to determine without cluttering the original

Reviewer 3: ```suggestion
2. Execution order of the rollback migration is not easy to determine without cluttering the original
```

migration and

> 3. wrapping migration execution in a transaction makes rollback obsolete, if you are working with a database

Reviewer 3: ```suggestion
3. Wrapping migration execution in a transaction makes rollback obsolete, if you are working with a database
```

that supports transactional DDL.

Having no support for rollbacks conducts another nice effect: our migration directory consists of a bunch of simple SQL

> files, each of them standing for a version of the database schema. Atlas migration filename format follows a very simple

Reviewer 3: ```suggestion
files, each of them expressing a version of the database schema. Atlas migration filename format follows a very simple
```

Not sure 'expressing' is necessarily the better term, but I'd say something like that or 'being' or 'each one being a version of the database schema.'

structure: `version_[name].sql`, with the `name` being optional. `version` can be an arbitrary string. Migration files
are lexicographically sorted by filename.

<Tabs
    defaultValue="tree"
    values={[
        {label: 'Tree', value: 'tree'},
        {label: '1_initial.sql', value: '1'},
        {label: '2_second.sql', value: '2'},
        {label: '3_second.sql', value: '3'},
        {label: 'atlas.sum', value: 'sum'},
]}>
<TabItem value="tree">

```shell
↪ tree .
.
├── 1_initial.sql
├── 2_second.sql
├── 3_third.sql
└── atlas.sum

0 directories, 4 files
```

</TabItem>

<TabItem value="1">

```sql
-- add new schema named "my_schema"
CREATE DATABASE `my_schema`;
-- create "tbl" table

> CREATE TABLE `my_schema`.`tbl` (`col` int NOT NULL);

Author: I am not sure, if I want to ad more value here like creating `users` and `pets` instead of simple `tbl`s and `col`s.

```

</TabItem>

<TabItem value="2">

```sql
ALTER TABLE `my_schema`.`tbl` ADD `col_2` TEXT;
```

</TabItem>

<TabItem value="3">

```sql
CREATE TABLE `tbl_2` (`col` int NOT NULL);
```

</TabItem>

<TabItem value="sum">

```text
h1:cD9kOv5VDRLrKVZ0pM4CxAlhH6mgE8PQLpUeuIMKDcs=
1_initial.sql h1:SrFyOe0eg5WnE96GH3TtAt6046sfrOK4YKZYBlYr1SA=
2_second.sql h1:FPzwV+MzwyCss7SASZtyafXiYc9Un5bzlcc3u7MxLJU=
3_third.sql h1:uD+xDcA3Q+gHqwca2ZBDAHWYvC2eiUcwr1IgMsN0Q6c=

```

</TabItem>

</Tabs>

If you want to follow along, you can simply copy and paste the above files in a folder on your system. Make sure you
have a database ready to work on. You can start an ephemeral docker container with the following command:

```shell
# Run a local mysql container listening on port 3306.

> docker run --rm --name atlas-apply --detach --env MYSQL_ROOT_PASSWORD=pass -p 3306:3306 mysql

Reviewer 2: for "future-proofness" , use a specific version tag for the container

```

### Apply Migrations

In order to apply migrations you need to have the Atlas CLI in version v0.7.0 or above. Follow the
[installation instructions](https://atlasgo.io/getting-started#installation) if you have no version installed yet.

Now, to apply the first migration of our migration directory, we call `atlas migrate apply` and pass in some
configuration parameters.

```shell
atlas migrate apply 1 \
  --dir "file://migrations" \
  --url "mysql://root:pass@localhost:3306/"
```

```shell
Migrating to version 1 (1 migrations in total):

  -- migrating version 1
    -> CREATE DATABASE `my_schema`;
    -> CREATE TABLE `my_schema`.`tbl` (`col` int NOT NULL);
  -- ok (17.247319ms)

  -------------------------
  -- 18.784204ms
  -- 1 migrations
  -- 2 sql statements

```

### Migration Status

Atlas saves information about the database schema revisions (applied migration versions) in a special table called

> `atlas_schema_revisions`. Since in our example above the connection was schema-less (we call that `Database Realm`),
> Atlas put the revision table into its own schema called `atlas_schema_revisions` as well. For a schema bound connection

Reviewer 2: IMO introducing the internal concept of a Realm isn't useful for this discussion
```suggestion
`atlas_schema_revisions`. In the example above we connected to the database without specifying which schema to operate against.
For this reason, Atlas created the revision table in a new schema called `atlas_schema_revisions`. For a schema bound connection
```

Atlas will put the table into the connected schema. We will see that in a bit.

Go ahead and call `atlas migrate status` to gather information about the database migration state:

```shell
atlas migrate status \
  --dir "file://migrations" \
  --url "mysql://root:pass@localhost:3306/"
```

```shell
Migration Status: PENDING
  -- Current Version:  1
  -- Next Version:     2
  -- Executed Files:   1
  -- Pending Files:    2

```

This output tells us, that the last applied version is `1`, the next one is called `2` and that we still have 2
migrations pending. Let's apply the pending migrations:

> Note, that we pass in no argument to the `apply` in which case Atlas will attempt to apply all pending migrations.

Reviewer 3: ```suggestion
Note, that we do not pass an argument to the `apply`, in which case Atlas will attempt to apply all pending migrations.
```

```shell
atlas migrate apply \
  --dir "file://migrations" \
  --url "mysql://root:pass@localhost:3306/"
```

```shell
Migrating to version 3 from 1 (2 migrations in total):

  -- migrating version 2
    -> ALTER TABLE `my_schema`.`tbl` ADD `col_2` TEXT;
  -- ok (13.98847ms)

  -- migrating version 3
    -> CREATE TABLE `tbl_2` (`col` int NOT NULL);
     Error 1046: No database selected

  -------------------------
  -- 15.604338ms
  -- 1 migrations ok (1 with errors)
  -- 1 sql statements ok (1 with errors)

Error: Execution had errors:
Error 1046: No database selected

Error: sql/migrate: execute: executing statement "CREATE TABLE `tbl_2` (`col` int NOT NULL);" from version "3": Error 1046: No database selected
exit status 1

```

> What did happen there? After further investigation, you will find that our connection URL is bound to the realm, not to

Reviewer 3: ```suggestion
What happened here? After further investigation, you will find that our connection URL is bound to the realm, not to
```

> What did happen there? After further investigation, you will find that our connection URL is bound to the realm, not to

Reviewer 2: ```suggestion
What did happen there? After further investigation, you will find that our connection URL is bound to the entire database, not to
```

a schema. The third migration file however does not contain a schema qualifier for the `CREATE TABLE` statement.

By default, Atlas wraps the execution of each migration file into one transaction. This transaction gets rolled back
if any error occurs withing execution. Be aware though, that some databases don't support transactional DDL (e.g.

> MySQL / MariaDB). if you want to learn how to configure the transaction behavior of Atlas, have a look at the

Reviewer 3: ```suggestion
MySQL / MariaDB). If you want to learn how to configure the transaction behavior of Atlas, have a look at the
```

> if any error occurs withing execution. Be aware though, that some databases don't support transactional DDL (e.g.
> MySQL / MariaDB). if you want to learn how to configure the transaction behavior of Atlas, have a look at the

Reviewer 2: ```suggestion
if any error occurs withing execution. Be aware though, that some databases, such as MySQL and MariaDB, don't support transactional DDL. If you want to learn more about how to configure the way Atlas uses transactions, have a look at the
```

> [docs](/versioned/apply).

Reviewer 2: ```suggestion
[docs](/versioned/apply#transaction-configuration).
```

### Migration Retry

> The first solution would be to add the migration file and add a qualifier to the statement, so it looks like below:

Reviewer 3: ```suggestion
The first solution would be to add the migration file and add a qualifier to the statement:
```

> The first solution would be to add the migration file and add a qualifier to the statement, so it looks like below:

Reviewer 2: Is there a second solution?  Otherwise, "To resolve this.."

```sql
CREATE TABLE `my_schema`.`tbl_2` (`col` int NOT NULL);
```

Since you changed the contents of a migration file we have to re-calculate the directory integrity hash-sum by calling:

```shell
atlas migrate hash --force \
  --dir "file://migrations"
```

We can then proceed and simply attempt to execute the migration file again.

```shell
atlas migrate apply \
  --dir "file://migrations" \
  --url "mysql://root:pass@localhost:3306/"
```

```shell
Migrating to version 3 from 2 (1 migrations in total):

  -- migrating version 3
    -> CREATE TABLE `my_schema`.`tbl_2` (`col` int NOT NULL);
  -- ok (15.168892ms)

  -------------------------
  -- 16.741173ms
  -- 1 migrations
  -- 1 sql statements

```

> Attempting to migrate again or calling `atlas migrate status` will tell us, that all migrations have been applied onto

Reviewer 3: ```suggestion
Attempting to migrate again or calling `atlas migrate status` will tell us that all migrations have been applied onto
```

the database and there is nothing to do at the moment.

<Tabs
    defaultValue="apply"
    values={[
        {label: 'atlas migrate apply', value: 'apply'},
        {label: 'atlas migrate status', value: 'status'},
]}>
<TabItem value="apply">

```shell
atlas migrate apply \
  --dir "file://migrations" \
  --url "mysql://root:pass@localhost:3306/"
```

```shell
The migration directory is synced with the database, no migration files to execute
```

</TabItem>

<TabItem value="status">

```shell
atlas migrate status \
  --dir "file://migrations" \
  --url "mysql://root:pass@localhost:3306/"
```

```shell
Migration Status: OK
  -- Current Version:  3
  -- Next Version:     Already at latest version
  -- Executed Files:   3
  -- Pending Files:    0

```

</TabItem>

</Tabs>

### Baseline

> There is a good chance someone wants to move to Atlas for an existing project where a database schema does already

Reviewer 3: ```suggestion
There is a good chance someone wants use Atlas on an existing project where a database schema already
```

> exist. In this case one could create an initial migration file reflecting the current state of a database schema by

Reviewer 3: ```suggestion
exists. In this case, one could create an initial migration file reflecting the current state of a database schema by
```

> ### Baseline
> 
> There is a good chance someone wants to move to Atlas for an existing project where a database schema does already
> exist. In this case one could create an initial migration file reflecting the current state of a database schema by

Reviewer 2: ```suggestion
### Moving an existing project to Atlas with Baseline Migrations

Another common scenario is when you need to move an existing project to Atlas. To do so, create an initial migration file reflecting the current state of a database schema by
```

> using `atlas migrate diff`. A very simple demonstration for this post is to head over to the database from before,

Reviewer 3: ```suggestion
using `atlas migrate diff`. A very simple way to do so would be by heading over to the database from before,
```

> delete the `atlas_schema_revisions` schema, empty your migration directory and run the `atlas migrate diff` command.

Reviewer 3: ```suggestion
deleting the `atlas_schema_revisions` schema, emptying the migration directory and running the `atlas migrate diff` command.
```

```shell
rm -rf migrations
mysql -h 27.0.0.1 -uroot -ppass -e "CREATE SCHEMA `my_schema_dev`;" # create a dev-db
mysql -h 127.0.0.1 -uroot -ppass -e "DROP SCHEMA `atlas_schema_revisions`;"
atlas migrate diff \
  --dir "file://migrations" \
  --to "mysql://root:pass@localhost:3306/my_schema" \
  --dev-url "mysql://root:pass@localhost:3306/my_schema_dev"
```

> To demonstrate, that Atlas can also work on a schema level instead of a realm connection, we are running the on a

Reviewer 3: ```suggestion
To demonstrate that Atlas can also work on a schema level instead of a realm connection, we are running on a
```

```

</TabItem>

<TabItem value="sum">

```text
h1:5zAQ8J0qziTKWg09fRNdbUf3rnLdvA1KHayh8l1SgM0=
20220908105652.sql h1:4WEB39tqALVYNQJTfULKizxEuUC37sgFs0LN5dKJpOw=

```

</TabItem>

</Tabs>

Now, let's create a new migration file to create a table `tbl_3` and update the directory integrity file.

```shell
atlas migrate new add_table --dir "file://migrations"
echo "CREATE TABLE `tbl_3` (`col` text NULL);" >> migrations/$(ls -t migrations | head -n1)
atlas migrate hash --force --dir "file://migrations"
```

Since we now have both a migration file representing our current database state and the new migration file to apply,
we can make use of the `--baseline` flag:

```shell
atlas migrate apply \
  --dir "file://migrations" \
  --url "mysql://root:pass@localhost:3306/my_schema" \
  --baseline "20220908110527" # replace the version with the one generated by you
```

```shell
Migrating to version 20220908110847 from 20220908110527 (1 migrations in total):

  -- migrating version 20220908110847
    -> CREATE TABLE `tbl_3` (`col` text NULL);
  -- ok (14.325493ms)

  -------------------------
  -- 15.786455ms
  -- 1 migrations
  -- 1 sql statements
```

### Outlook

> The Atlas migration execution engine is used already within Ariga for several months. We will continue working on

Reviewer 3: ```suggestion
The Atlas migration execution engine is already being used by Ariga for several months. We will continue working on
```

> improving it, and you can expect many cool feature, like assisted troubleshooting for failed migrations, a more

Reviewer 3: ```suggestion
improving it, and releasing cool features, such as assisted troubleshooting for failed migrations, and more
```

> intelligent dialect aware execution planning for things like MySQL implicits commits and more.

Reviewer 3: ```suggestion
intelligent dialect-aware execution planning for things like MySQL implicits commits and more.
```
